Webhook notifications for receive mode

micahflee opened the pull request:
Resolves #1322.

This lets you optionally set a webhook URL. Whenever someone uploads something to your receive mode service, it makes a POST request to this URL (over Tor, unless you're running with `--local-only`) saying the number of files that were uploaded.

The simplest way to test is to use the keybase webhookbot.

## Create a webhook URL

For example, you can tell it `!webhook create onionshare-test` to generate a webhook URL:

![Screenshot from 2021-04-11 21-25-53](https://user-images.githubusercontent.com/156128/114340494-85b3d480-9b0c-11eb-9407-a98674bda0be.png)

## Test in the CLI

```
$ poetry run onionshare-cli --receive --webhook-url https://bots.keybase.io/webhookbot/Ieah6SBHq5fYhA7ZqjoryQWx4y0
```

Load the URL in Tor Browser and upload a few files. In my case, I uploaded 5. Then in keybase, I get this message from webhookbot:

![Screenshot from 2021-04-11 21-29-13](https://user-images.githubusercontent.com/156128/114340704-fc50d200-9b0c-11eb-98bc-1156ef9ac5aa.png)

## Test in the GUI

Open receive mode and check "Use webhook URL":

![Screenshot from 2021-04-11 21-30-10](https://user-images.githubusercontent.com/156128/114340765-21dddb80-9b0d-11eb-8c9c-66a9f5f52c96.png)

It has placeholder text:

![Screenshot from 2021-04-11 21-30-38](https://user-images.githubusercontent.com/156128/114340791-2efaca80-9b0d-11eb-9711-ee4dd5533bac.png)

Add the keybase URL, start the server, load it in Tor Browser, upload some files, and get the notification again.

micahflee commented:
One thing I didn't do is write a test for this. To test it, I'd basically need to start a web server, set it as the webhook URL, and wait for a POST request to come in when I download files, and this seems challenging to do in a test. I'd have to like, maybe run a flask service in a separate thread or something?

It might be worth it to add a test though anyway...

SaptakS commented:
Can't we mock a webhook URL?So that once it receives a request when a file is uploaded it print some text or something? I can see it will be complicated, but I would prefer mocking rather than running a flask service for testing purpose if possible.

SaptakS reviewed (commented):
Tested it via CLI and desktop, looks good to me. My only concern is the notification message and the label for the webhook might not be too verbose.

I would probably prefer "notification webhook URL" for desktop label. Otherwise it's not clear webhook for what exactly.
And in the message something more verbose like "x files uploaded to your onionshare receive instance". Right now if someone doesn't keep a proper name for their bot, it might confuse as to where the file is uploaded.

micahflee commented:
I fixed a bug where if you unchecked the webhook checkbox, it didn't actually unset the webhook. It does now.

I added a CLI test for webhooks (and also for uploading files at all in receive mode -- that wasn't being tested in CLI, only in GUI).

Now the webhook URL label says "Use notification webhook":

![Screenshot from 2021-04-12 09-36-18](https://user-images.githubusercontent.com/156128/114429770-972fc780-9b72-11eb-8498-906a0fb91b6f.png)

And the actual webhook message itself is more verbose too, like:

> 1 file uploaded to OnionShare

or

> 4 files uploaded to OnionShare

Ready for a final review I think.

SaptakS commented on cli/tests/test_cli_web.py:
```diff
@@ -12,6 +13,21 @@
+    webhook_data = data
+
+
+onionshare_cli.web.receive_mode.requests.post = requests_post_stub
```
This is great!

SaptakS reviewed (approved):
Looks good to me
